Add support for Windows to resource sys_info

mhackethal opened the pull request:
<!--- Provide a short summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail, what problems does it solve? -->
The resource sys_info was designed to support Linux and Windows OS.
The latest changes for (hostname) didn´t include the Windows support.
On Windows (OS) it Always returns that hostname is not supported for the OS yet.

In regards to #4570 i tried to add some of the Features for Windows.
This pull - is just a first guess. Maybe its better to re-design the whole resource to support
some kind of "initialize" method and have all data directly available?

Since most of the Windows "sys_info" are stored in the Win32_ComputerSystem i decided to create a "Cache" that holds all Information, instead of running a single query for every Attribute.

In case the functionality is welcome please provide me some hints how it should look like to fit into inspec. I will take care of These changes.

If its okay for you - the documentation and test updates will follow as soon as we have an allignment ?

## Related Issue
<!--- If you are suggesting a new feature or change, please create an issue first -->
<!--- Please link to the issue, discourse, or stackoverflow here: -->
#4570 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [x] Breaking change (a content change which would break existing functionality or processes)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.

mhackethal committed:
Update sys_info.rb

mhackethal committed:
Update sys_info to support some windows features.

Signed-off-by: mhackethal <mh@it31.de>

miah commented on lib/inspec/resources/sys_info.rb:
```diff
@@ -106,11 +158,29 @@ def model
       elsif os.windows?
-        inspec.powershell("Get-CimInstance -ClassName Win32_ComputerSystem | Select Model -ExpandProperty Model").stdout.chomp
+        if @cache.nil?
+          @cache = win_cache
```
Instead of this conditional you can do:

`@cache ||= win_cache`

I'd look for similar chunks of code in this PR and see if you can apply similar logic. I'd also have one thing that sets @cache and just have everything else that needs the value for @cache to call it,

eg;

```
def cache
  @cache ||= win_cache
end
```

miah commented on lib/inspec/resources/sys_info.rb:
```diff
@@ -34,15 +52,46 @@ def hostname(opt = nil)
+      if @cache.nil?
+        @cache = win_cache
+      end
+      if !opt.nil?
```
`unless opt.nil?` is nicer than the negation. But since we're just checking that opt is set, `if opt` should suffice and the `else` can cover the case where its nil.

mhackethal committed:
re-structure of sys_info

rename calss from system to sysinfo
split between windos + linux
harmonize output for windows and linux
Signed-off-by: mhackethal <mh@it31.de>

mhackethal committed:
re-structure of sys_info

rename calss from system to sysinfo
split between windos + linux
harmonize output for windows and linux
Signed-off-by: mhackethal <mh@it31.de>
fix whitespaces

clintoncwolfe commented on lib/inspec/resources/sys_info.rb:
```diff
@@ -3,117 +3,140 @@
-        linux_hostname(opt)
-      elsif os.darwin?
-        mac_hostname(opt)
+        @cache = LinuxInfo.new(inspec)
```
While you have a ivar named `@cache`, it actually isn't being used as a cache - you're repeatedly calling `info()` on it, which is expensive. The return value of info is what you want to cache.

Consider modifying this to:
```
@cache = LinuxInfo.new(inspec).info
```

and then (for example)

```
    def domain
      @cache[:domain]
    end
```

clintoncwolfe commented on lib/inspec/resources/sys_info.rb:
```diff
@@ -3,117 +3,140 @@
     end
   end
+
+  class LinuxInfo < SysInfo
```
I see a LinuxInfo and a new WindowsInfo, but the existing functionality supports MacOS as well. We'll need a support class for Mac.

clintoncwolfe commented on lib/inspec/resources/sys_info.rb:
```diff
@@ -3,117 +3,140 @@
+        domain = info["Domain"]
+      end
+
+      memory = info["TotalPhysicalMemory"] / 1024 rescue nil
```
I'd prefer `info.fetch("TotalPhysicalMemory", 0) / 1024` to resorting to exception handling here.

mhackethal replied:
sorry for my stupid Questions. 
Will info.fetch("TotalPhysicalMemory",0) / 1024 not raise an divion by Zero error?
In case TotalPhysicalMemory can not be found as key it will Reply the value as 0 and then try to divide it through 1024?
Or do i missunderstand this function?

zenspider replied:
0/1 is always 0. 1/0 is undefined/error.

mhackethal replied:
yah my fault. already figured that out ;)

clintoncwolfe reviewed (changes requested):
Thanks for the contribution, and for coming back to this!

Looks like we're missing some mac functionality, and we'll definitely need tests on this. A couple of other minor code changes requested. Thanks!

mhackethal committed:
re-add Darwin support

Change Cache as mentioned by @clintonwolfe
replace Rescue Nil with info.fetch("TotalPhysicalMemory",0)
Signed-off-by: mhackethal <mh@it31.de>

mhackethal committed:
update to hold status-quo

allow the usage of sys_info.hostname('opt')
obvious fixes for whitespaces
Signed-off-by: mhackethal <mh@it31.de>

mhackethal committed:
Update sys_info.md.erb

obvious fix: ip-address should ip_address
Add totalmemory_kb

miah commented on lib/inspec/resources/sys_info.rb:
```diff
@@ -3,117 +3,182 @@
+        manufacturer: manufacturer,
+        model: model,
+        totalmemory_kb: totalmemory_kb,
+      }
```
Can we construct this structure higher? Doing this in each SysInfo subclass is likely to cause issues when the structure changes in one, but not the other.

mhackethal replied:
I think i got your point but moving this into the sys_info class would cause massives changes and will result in different layout as all other ressources: like "services","package". 
But i think i got your point. 
What about updating the methods in root class "sys_info" in this way.

```
    def foobar
      @cache[:foobar] || "The `sys_info.foobar` resource is not supported on your OS yet."
    end
```

In this case the method will return "unsupported" in case the information is not available by the OS subclass.
If this approach is followed by the next changes it will ensure that you can have different methods foreach OS. If a method is not available on a specific OS inspec will notify you.

miah commented on lib/inspec/resources/sys_info.rb:
```diff
@@ -3,117 +3,182 @@
+      short = inspec.command("hostname -s").stdout.chomp
+      manufacturer = inspec.command("cat /sys/class/dmi/id/sys_vendor").stdout.chomp
+      model = inspec.command("cat /sys/class/dmi/id/product_name").stdout.chomp
+      totalmemory_kb = inspec.command("awk '/MemTotal/ {print $2}' /proc/meminfo").stdout.chomp
```
I would probably create these as methods in the SysInfo subclasses, we'll get lots of bonus usability.

```
linux = LinuxInfo.new(...)
foo.fqdn
foo.ip
```

mhackethal replied:
Here , i dont get your point. The resource have the methods within "sys_info" what is the advantage to have a copy of that foreach OS?
The subclasses foreach OS is only available inside sys_info ressource so its not usable directly inside the inspec tests.
Even if i change it in that way i would have to create a complex way of execution and caching for the windows side.
Or i have to split the 2 windows commands into 6 single powershell commands where i think this even dont bring any advantage.

miah reviewed (changes requested):
Thanks for the contribution, just a few changes I'd like to see.

mhackethal committed:
minor changes and improvements

Improve return values if method is not supported by OS
Replace rescue nil at division with fetch statement.
Remove no longer required information from darwin class
Signed-off-by: mhackethal <mh@it31.de>
